src/device_trezor/device_trezor.cpp

```diff
@@ -0,0 +1,354 @@
+      if (!trezor_device) {
+        trezor_device = new device_trezor();
+        trezor_device->set_name(HW_TREZOR_NAME);
+      }
+      return trezor_device;
+    }
+
+#if WITH_DEVICE_TREZOR_LITE
+    static device_trezor_lite *ensure_trezor_device_lite(){
+      if (!trezor_device_lite) {
+        trezor_device_lite = new device_trezor_lite();
+        trezor_device_lite->set_name(HW_TREZOR_NAME);
+      }
+      return trezor_device_lite;
+    }
+#endif
+
+    void register_all(std::map<std::string, std::unique_ptr<device>> &registry) {
+      registry.insert(std::make_pair(HW_TREZOR_NAME, std::unique_ptr<device>(ensure_trezor_device())));
+#if WITH_DEVICE_TREZOR_LITE
+      registry.insert(std::make_pair(HW_TREZOR_NAME_LITE, std::unique_ptr<device>(ensure_trezor_device_lite())));
+#endif
+    }
+
+    void register_all() {
+      hw::register_device(HW_TREZOR_NAME, ensure_trezor_device());
+#if WITH_DEVICE_TREZOR_LITE
+      hw::register_device(HW_TREZOR_NAME_LITE, ensure_trezor_device_lite());
+#endif
+    }
+
+    device_trezor::device_trezor() {
+
+    }
+
+    device_trezor::~device_trezor() {
+      try {
+        disconnect();
+        release();
+      } catch(std::exception & e){
+        LOG_PRINT_L1(std::string("Could not disconnect and release: ") + e.what());
+      }
+    }
+
+    /* ======================================================================= */
+    /*                              SETUP/TEARDOWN                             */
+    /* ======================================================================= */
+
+    /* ======================================================================= */
+    /*  LOCKER                                                                 */
+    /* ======================================================================= */
+
+    //automatic lock one more level on device ensuring the current thread is allowed to use it
+    #define AUTO_LOCK_CMD() \
+      /* lock both mutexes without deadlock*/ \
+      boost::lock(device_locker, command_locker); \
+      /* make sure both already-locked mutexes are unlocked at the end of scope */ \
+      boost::lock_guard<boost::recursive_mutex> lock1(device_locker, boost::adopt_lock); \
+      boost::lock_guard<boost::mutex> lock2(command_locker, boost::adopt_lock)
+
+    /* ======================================================================= */
+    /*                             WALLET & ADDRESS                            */
+    /* ======================================================================= */
+
+    bool device_trezor::get_public_address(cryptonote::account_public_address &pubkey) {
+      auto res = get_address();
+
+      cryptonote::address_parse_info info;
+      bool r = cryptonote::get_account_address_from_str(info, this->network_type, res->address());
+      if (!r){
+        LOG_PRINT_L2("Returned address parse fail: " + res->address());
+        throw std::runtime_error("Could not parse returned address");
```

Comment:
This function returns a bool, but also throws, is this intended ?
There's also a useful CHECK_AND_ASSERT_MES macro which you might find useful for these cases. It check, logs and returns on check failure.

Reply:
Up to you. If you choose exceptions, return void though, otherwise it's confusing.

Reply:
I mean, I would have to modify also the Leger code as this is a common function.

Reply:
Typical example is `void account_base::create_from_device(hw::device &hwdev)` which does not check return codes. 

```
hwdev.connect();
hwdev.get_public_address(m_keys.m_account_address);
hwdev.get_secret_keys(m_keys.m_view_secret_key, m_keys.m_spend_secret_key);
```

They return bool.